File: typescript/src/observability.ts
/**
 * Metric data for a single node execution.
 */
export interface ExecutionMetric {
  runId: string;
  nodeId: string;
  nodeName: string;
  status: string;
  durationMs: number;
  errorType?: string;
  errorCode?: string;
  tokenUsage: number;
  timestamp: number;
}

/**
 * Creates an execution metric with current timestamp.
 * @param opts - Metric data without timestamp
 * @returns Complete execution metric
 */
export function createExecutionMetric(opts: Omit<ExecutionMetric, "timestamp">): ExecutionMetric {
  return { ...opts, timestamp: Date.now() / 1000 };
}

/**
 * Converts an execution metric to a dictionary.
 * @param metric - Execution metric to convert
 * @returns Dictionary with snake_case keys
 */
export function metricToDict(metric: ExecutionMetric): Record<string, unknown> {
  return {
    run_id: metric.runId,
    node_id: metric.nodeId,
    node_name: metric.nodeName,
    status: metric.status,
    duration_ms: metric.durationMs,
    error_type: metric.errorType ?? null,
    error_code: metric.errorCode ?? null,
    token_usage: metric.tokenUsage,
    timestamp: metric.timestamp,
  };
}

/**
 * Collects and aggregates execution metrics.
 */
export class MetricsCollector {
  private _metrics: ExecutionMetric[] = [];
  private _maxRecords: number;

  /**
   * Creates a new metrics collector.
   * @param maxRecords - Maximum number of metrics to store (default: 10000)
   */
  constructor(maxRecords = 10000) {
    this._maxRecords = maxRecords;
  }

  /**
   * Records an execution metric.
   * @param metric - Execution metric to record
   */
  record(metric: ExecutionMetric): void {
    this._metrics.push(metric);
    if (this._metrics.length > this._maxRecords) {
      this._metrics = this._metrics.slice(-this._maxRecords);
    }
  }

  /**
   * Gets summary statistics for recorded metrics.
   * @param lastN - Only include last N metrics (optional)
   * @returns Summary statistics dictionary
   */
  getSummary(lastN?: number): Record<string, unknown> {
    const metrics = lastN ? this._metrics.slice(-lastN) : this._metrics;
    if (metrics.length === 0) {
      return { total_executions: 0 };
    }

    const passCount = metrics.filter((m) => m.status === "pass").length;
    const failCount = metrics.filter((m) => m.status === "fail").length;
    const durations = metrics.map((m) => m.durationMs);

    return {
      total_executions: metrics.length,
      pass_count: passCount,
      fail_count: failCount,
      success_rate: passCount / metrics.length,
      avg_duration_ms: durations.reduce((a, b) => a + b, 0) / durations.length,
      min_duration_ms: durations.reduce((a, b) => Math.min(a, b)),
      max_duration_ms: durations.reduce((a, b) => Math.max(a, b)),
      total_token_usage: metrics.reduce((sum, m) => sum + m.tokenUsage, 0),
    };
  }

  /**
   * Clears all collected metrics.
   */
  clear(): void {
    this._metrics = [];
  }
}

File: python/canvastekk_workflow_sdk/observability.py
"""
Metrics & Observability

Opt-in observability support for node execution.
Provides structured metrics collection without hard dependencies on
any specific observability backend.
"""

from __future__ import annotations

import logging
import threading
import time
from dataclasses import dataclass, field
from typing import Any

logger = logging.getLogger(__name__)


@dataclass
class ExecutionMetric:
    """A single execution metric record."""

    run_id: str
    node_id: str
    node_name: str
    status: str
    duration_ms: int
    error_type: str | None = None
    error_code: str | None = None
    token_usage: float = 0.0
    timestamp: float = field(default_factory=time.time)

    def to_dict(self) -> dict[str, Any]:
        """Serialize the metric to a JSON-friendly dictionary."""
        return {
            "run_id": self.run_id,
            "node_id": self.node_id,
            "node_name": self.node_name,
            "status": self.status,
            "duration_ms": self.duration_ms,
            "error_type": self.error_type,
            "error_code": self.error_code,
            "token_usage": self.token_usage,
            "timestamp": self.timestamp,
        }


class MetricsCollector:
    """Collects execution metrics in-process.

    Usage:
        collector = MetricsCollector()
        collector.record(metric)
        summary = collector.get_summary()
    """

    def __init__(self, max_records: int = 10000) -> None:
        """Initialize metrics collector.

        Args:
            max_records: Maximum number of metrics to keep (evicts oldest).
        """
        self._metrics: list[ExecutionMetric] = []
        self._max_records = max_records
        self._lock = threading.Lock()

    def record(self, metric: ExecutionMetric) -> None:
        """Append a metric, evicting the oldest entry when capacity is reached."""
        with self._lock:
            self._metrics.append(metric)
            if len(self._metrics) > self._max_records:
                self._metrics = self._metrics[-self._max_records :]

    def get_summary(self, last_n: int | None = None) -> dict[str, Any]:
        """Return aggregated statistics over the collected metrics.

        Args:
            last_n: If set, only consider the most recent *last_n* records.

        Returns:
            A dict with keys ``total_executions``, ``pass_count``,
            ``fail_count``, ``success_rate``, ``avg_duration_ms``,
            ``min_duration_ms``, ``max_duration_ms``, and
            ``total_token_usage``.  Returns ``{"total_executions": 0}``
            when no metrics have been recorded.
        """
        with self._lock:
            metrics = list(self._metrics[-last_n:]) if last_n else list(self._metrics)
        if not metrics:
            return {"total_executions": 0}

        pass_count = sum(1 for m in metrics if m.status == "pass")
        fail_count = sum(1 for m in metrics if m.status == "fail")
        durations = [m.duration_ms for m in metrics]

        return {
            "total_executions": len(metrics),
            "pass_count": pass_count,
            "fail_count": fail_count,
            "success_rate": pass_count / len(metrics) if metrics else 0.0,
            "avg_duration_ms": sum(durations) / len(durations),
            "min_duration_ms": min(durations),
            "max_duration_ms": max(durations),
            "total_token_usage": sum(m.token_usage for m in metrics),
        }

    def clear(self) -> None:
        """Remove all collected metrics."""
        with self._lock:
            self._metrics.clear()
